File: dotnet/ChunkL/Diagnostics/Diagnostic.cs
namespace ChunkL.Diagnostics;

/// <summary>
/// A single diagnostic message produced during parsing.
/// </summary>
public sealed class Diagnostic
{
    public DiagnosticSeverity Severity { get; }
    public string Message { get; }
    public SourcePosition Position { get; }
    public string? Code { get; }

    public Diagnostic(DiagnosticSeverity severity, string message, SourcePosition position, string? code = null)
    {
        Severity = severity;
        Message = message;
        Position = position;
        Code = code;
    }

    public override string ToString() =>
        $"{Severity} {Position}: {Message}" + (Code is not null ? $" [{Code}]" : "");
}

File: chunkl-rs/src/diagnostic.rs
use std::fmt;

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct SourcePosition {
    pub line: usize,
    pub column: usize,
}

impl SourcePosition {
    pub const fn new(line: usize, column: usize) -> Self {
        Self { line, column }
    }
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct SourceRange {
    pub start: SourcePosition,
    pub end: SourcePosition,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DiagnosticSeverity {
    Info,
    Warning,
    Error,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Diagnostic {
    pub severity: DiagnosticSeverity,
    pub message: String,
    pub position: SourcePosition,
    pub code: Option<String>,
}

impl Diagnostic {
    pub fn error(message: impl Into<String>, position: SourcePosition) -> Self {
        Self {
            severity: DiagnosticSeverity::Error,
            message: message.into(),
            position,
            code: None,
        }
    }
}

impl fmt::Display for Diagnostic {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            formatter,
            "{:?} ({},{}): {}",
            self.severity, self.position.line, self.position.column, self.message
        )?;
        if let Some(code) = &self.code {
            write!(formatter, " [{code}]")?;
        }
        Ok(())
    }
}
